composeApp/src/commonMain/kotlin/sp/bvantur/tasky/register/domain/RegisterUserUseCase.kt

```diff
@@ -0,0 +1,11 @@
+package sp.bvantur.tasky.register.domain
+
+import sp.bvantur.tasky.register.data.RegisterRepository
+
+class RegisterUserUseCase(private val repository: RegisterRepository) {
+    suspend operator fun invoke(name: String, email: String, password: String): Result<Unit> = repository.register(
```

Comment:
I have a question on how do you usually handle network response in your projects. 
1. What do you usually use to communicate results between layers? Kotlin Result class or something else?
2. To which layer do you then use that class for communicating the result of backend communication? Are you doing any mapping in the domain layer that prepares results for a presentation layer?
3. This question is optional and it depends probably on the answers from the first 2 questions, but how do you usually handle different error cases for backend communication? If you are using Kotlin Result capability, does it make sense to have different custom Exceptions created for this case? Or completely map the whole error case to a domain-specific result class and then work with that class in ViewModel?

Reply:
1. For a generic error handling wrapper, I love this version:
```
sealed interface Result<out D, out E: Error> {
    data class Success<out D>(val data: D): Result<D, Nothing>
    data class Error<out E: com.plcoding.tasky.core.domain.util.Error>(val error: E): Result<Nothing, E>
}

inline fun <T, E: Error, R> Result<T, E>.map(map: (T) -> R): Result<R, E> {
    return when(this) {
        is Result.Error -> Result.Error(error)
        is Result.Success -> Result.Success(map(data))
    }
}

fun <T, E: Error> Result<T, E>.asEmptyDataResult(): EmptyResult<E> {
    return map {  }
}

inline fun <T, E: Error> Result<T, E>.onSuccess(action: (T) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}
inline fun <T, E: Error> Result<T, E>.onError(action: (E) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}

typealias EmptyResult<E> = Result<Unit, E>
```
It handles data and errors in a type-safe way. You'll need a marker interface to mark error enums:
```
interface Error
```
And then you can extend this across the codebase with different error types, e.g.:
```
sealed interface DataError: Error {
    enum class Remote: DataError {
        NOT_FOUND,
        NO_INTERNET,
        UNAUTHORIZED,
        ...
    }
    
    enum class Local: DataError {
        DISK_FULL,
        ...
    }
}
```
Works great:
```
repository.signUp(payload).onSuccess { responseData ->
    // Type-safe response data
}.onError { error: DataError.Remote -> 
    _state.update { it.copy(error = error) }
}
```
2. I typically map my data models to domain models and then use these in presentation just like that. Only if the UI model is strongly different than the domain model (e.g. if lots of numbers/dates from the domain model have to formatted to show in the UI), then I create a mapper in presentation that maps domain to UI model.
3. The Result class from above takes care of this. For every set of errors, you simply create an enum class that implements the Error marker interface, e.g. for password validation errors:
```
enum class PasswordValidationError: Error {
    TOO_SHORT, TOO_LONG, NO_UPPERCASE, NO_LOWERCASE, NO_SPECIAL_CHAR
}
```